File: packages/flutter/lib/src/core/chromatic_pulse.dart
import 'dart:math' as math;
import 'dart:ui';
import 'constants.dart';

class ChromaticPulseState {
  final Color color;
  final double opacity;
  final double progress;

  const ChromaticPulseState({
    required this.color,
    required this.opacity,
    required this.progress,
  });
}

/// Cubic bezier solver for CHROMATIC_EASE.
double solveCubicBezier(
  double t, [
  double x1 = 0.45,
  double y1 = 0.0,
  double x2 = 0.55,
  double y2 = 1.0,
]) {
  if (t <= 0) return 0.0;
  if (t >= 1) return 1.0;

  double u = t;
  for (int i = 0; i < 8; i++) {
    final oneMinusU = 1.0 - u;
    final x = 3 * oneMinusU * oneMinusU * u * x1 +
        3 * oneMinusU * u * u * x2 +
        u * u * u;
    final dx = 3 * oneMinusU * oneMinusU * x1 +
        6 * oneMinusU * u * (x2 - x1) +
        3 * u * u * (1 - x2);
    if (dx.abs() < 1e-6) break;
    u -= (x - t) / dx;
    u = u.clamp(0.0, 1.0);
  }

  final oneMinusU = 1.0 - u;
  return 3 * oneMinusU * oneMinusU * u * y1 +
      3 * oneMinusU * u * u * y2 +
      u * u * u;
}

/// Color linear interpolation.
Color lerpColor(Color a, Color b, double t) {
  final clampedT = t.clamp(0.0, 1.0);
  final r = (a.r + (b.r - a.r) * clampedT).clamp(0.0, 1.0);
  final g = (a.g + (b.g - a.g) * clampedT).clamp(0.0, 1.0);
  final bl = (a.b + (b.b - a.b) * clampedT).clamp(0.0, 1.0);
  final alpha = (a.a + (b.a - a.a) * clampedT).clamp(0.0, 1.0);
  return Color.from(alpha: alpha, red: r, green: g, blue: bl);
}

/// Chromatic Pulse color cycle engine in Dart.
class ChromaticPulse {
  List<Color> palette;
  int durationMs;
  bool reducedMotion;

  ChromaticPulse({
    required this.palette,
    this.durationMs = LoaderConstants.chromaticPulseDurationMs,
    this.reducedMotion = false,
  }) {
    if (palette.length < 2) {
      palette = [const Color(0xFF4A90E2), const Color(0xFF50E3C2)];
    }
  }

  void updateConfig({List<Color>? palette, int? durationMs, bool? reducedMotion}) {
    if (palette != null && palette.length >= 2) {
      this.palette = palette;
    }
    if (durationMs != null) this.durationMs = durationMs;
    if (reducedMotion != null) this.reducedMotion = reducedMotion;
  }

  ChromaticPulseState evaluateAt(double elapsedMs) {
    if (reducedMotion) {
      return ChromaticPulseState(
        color: palette[0],
        opacity: 1.0,
        progress: 0.0,
      );
    }

    final totalDuration = math.max(1, durationMs);
    final progress = (elapsedMs % totalDuration) / totalDuration;

    final count = palette.length;
    final scaledProgress = progress * count;
    final index = scaledProgress.floor();
    final nextIndex = (index + 1) % count;
    final segmentProgress = scaledProgress - index;

    final easedSegment = solveCubicBezier(
      segmentProgress,
      LoaderConstants.chromaticEaseBezier[0],
      LoaderConstants.chromaticEaseBezier[1],
      LoaderConstants.chromaticEaseBezier[2],
      LoaderConstants.chromaticEaseBezier[3],
    );

    final color = lerpColor(palette[index], palette[nextIndex], easedSegment);

    final breathingFactor = 0.5 * (1 + math.sin(progress * 2 * math.pi));
    final opacity = LoaderConstants.chromaticOpacityMin +
        (LoaderConstants.chromaticOpacityMax - LoaderConstants.chromaticOpacityMin) *
            breathingFactor;

    return ChromaticPulseState(
      color: color,
      opacity: opacity.clamp(0.0, 1.0),
      progress: progress,
    );
  }
}

File: packages/swift/Sources/B1CodesLoaders/Core/ChromaticPulse.swift
import SwiftUI

public struct ChromaticPulseState {
    public let color: Color
    public let opacity: Double
    public let progress: Double

    public init(color: Color, opacity: Double, progress: Double) {
        self.color = color
        self.opacity = opacity
        self.progress = progress
    }
}

/// Solves cubic bezier curve Y for progress t [0, 1].
public func solveCubicBezier(
    t: Double,
    x1: Double = 0.45,
    y1: Double = 0.0,
    x2: Double = 0.55,
    y2: Double = 1.0
) -> Double {
    if t <= 0 { return 0.0 }
    if t >= 1 { return 1.0 }

    var u = t
    for _ in 0..<8 {
        let oneMinusU = 1.0 - u
        let x = 3 * oneMinusU * oneMinusU * u * x1 + 3 * oneMinusU * u * u * x2 + u * u * u
        let dx = 3 * oneMinusU * oneMinusU * x1 + 6 * oneMinusU * u * (x2 - x1) + 3 * u * u * (1 - x2)
        if abs(dx) < 1e-6 { break }
        u -= (x - t) / dx
        u = max(0.0, min(1.0, u))
    }

    let oneMinusU = 1.0 - u
    return 3 * oneMinusU * oneMinusU * u * y1 + 3 * oneMinusU * u * u * y2 + u * u * u
}

/// Linear interpolation between two SwiftUI Colors.
public func lerpColor(from colorA: Color, to colorB: Color, fraction: Double) -> Color {
    let t = max(0.0, min(1.0, fraction))

    #if canImport(UIKit)
    let nativeA = UIColor(colorA)
    let nativeB = UIColor(colorB)
    #elseif canImport(AppKit)
    let nativeA = NSColor(colorA).usingColorSpace(.sRGB) ?? NSColor(colorA)
    let nativeB = NSColor(colorB).usingColorSpace(.sRGB) ?? NSColor(colorB)
    #endif

    var rA: CGFloat = 0, gA: CGFloat = 0, bA: CGFloat = 0, aA: CGFloat = 0
    var rB: CGFloat = 0, gB: CGFloat = 0, bB: CGFloat = 0, aB: CGFloat = 0

    nativeA.getRed(&rA, green: &gA, blue: &bA, alpha: &aA)
    nativeB.getRed(&rB, green: &gB, blue: &bB, alpha: &aB)

    let r = rA + (rB - rA) * CGFloat(t)
    let g = gA + (gB - gA) * CGFloat(t)
    let b = bA + (bB - bA) * CGFloat(t)
    let a = aA + (aB - aA) * CGFloat(t)

    return Color(red: Double(r), green: Double(g), blue: Double(b), opacity: Double(a))
}

/// Chromatic Pulse color cycle engine in Swift.
public class ChromaticPulse {
    public var palette: [Color]
    public var durationMs: Double
    public var reducedMotion: Bool

    public init(
        palette: [Color],
        durationMs: Double = LoaderConstants.chromaticPulseDurationMs,
        reducedMotion: Bool = false
    ) {
        self.palette = palette.count >= 2 ? palette : [Color.blue, Color.teal]
        self.durationMs = durationMs
        self.reducedMotion = reducedMotion
    }

    public func updateConfig(palette: [Color]? = nil, durationMs: Double? = nil, reducedMotion: Bool? = nil) {
        if let palette = palette, palette.count >= 2 {
            self.palette = palette
        }
        if let durationMs = durationMs { self.durationMs = durationMs }
        if let reducedMotion = reducedMotion { self.reducedMotion = reducedMotion }
    }

    public func evaluateAt(elapsedMs: Double) -> ChromaticPulseState {
        if reducedMotion {
            return ChromaticPulseState(
                color: palette[0],
                opacity: 1.0,
                progress: 0.0
            )
        }

        let totalDuration = max(1.0, durationMs)
        let progress = (elapsedMs.truncatingRemainder(dividingBy: totalDuration)) / totalDuration

        let count = palette.count
        let scaledProgress = progress * Double(count)
        let index = Int(floor(scaledProgress))
        let nextIndex = (index + 1) % count
        let segmentProgress = scaledProgress - Double(index)

        let easedSegment = solveCubicBezier(
            t: segmentProgress,
            x1: LoaderConstants.chromaticEaseBezier.x1,
            y1: LoaderConstants.chromaticEaseBezier.y1,
            x2: LoaderConstants.chromaticEaseBezier.x2,
            y2: LoaderConstants.chromaticEaseBezier.y2
        )

        let color = lerpColor(from: palette[index], to: palette[nextIndex], fraction: easedSegment)

        let breathingFactor = 0.5 * (1.0 + sin(progress * 2.0 * .pi))
        let opacity = LoaderConstants.chromaticOpacityMin +
            (LoaderConstants.chromaticOpacityMax - LoaderConstants.chromaticOpacityMin) * breathingFactor

        return ChromaticPulseState(
            color: color,
            opacity: max(0.0, min(1.0, opacity)),
            progress: progress
        )
    }
}
